File: built-in plugins dsp/mb-vocal-formant.cpp
/**
 * MB Formant Shifter
 * Category : effect
 * Type     : vocal
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Vocal formant manipulation
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_VOCAL_FORMANT_H
#define MB_VOCAL_FORMANT_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbVocalFormant : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-vocal-formant";
    static constexpr const char* PLUGIN_NAME    = "MB Formant Shifter";
    static constexpr const char* PLUGIN_TYPE    = "vocal";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float shift = 0f;  // range [-12, 12]
    float character = 0.5f;  // range [0, 1]
    };

    MbVocalFormant() = default;
    ~MbVocalFormant() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.shift = std::clamp(params.shift, -12f, 12f);
        params.character = std::clamp(params.character, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Formant Shifter
        return input;
    }
};

#endif // MB_VOCAL_FORMANT_H

File: maxbooster7.5/built-in plugins dsp/mb-vocal-formant.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbVocalFormantPlugin: PluginDefinition = {
  id: "mb-vocal-formant",
  slug: "mb-vocal-formant",
  name: "MB Formant Shifter",
  category: "effect",
  type: "vocal",
  version: "1.0.0",
  description: "Vocal formant manipulation",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "shift",
      name: "Formant Shift",
      type: "float",
      defaultValue: 0,
      minValue: -12,
      maxValue: 12,
      automatable: true,
    },
    {
      id: "character",
      name: "Character",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { shift: 0, character: 0.5 },
};

export default MbVocalFormantPlugin;
